src/Cms/Language.php

```diff
@@ -472,12 +474,19 @@ public function save(): static
 			'direction'    => $this->direction(),
 			'locale'       => Locale::export($this->locale()),
 			'name'         => $this->name(),
-			'translations' => $this->translations(),
+			'translations' => $this->translations()->toArray(),
 			'url'          => $this->url,
 		];
 
 		ksort($data);
 
+		// save translations to the custom root and remove translations
+		// to prevent duplication write into the language file
+		if ($this->translations()->root() !== null) {
+			$this->translations()->save($data['translations'] ?? []);
+			$data['translations'] = [];
+		}
```

Comment:
How about instead of unsetting it here, we do an if-else clause: The if part writes it to the dedicated root, the else part adds it to the `$data` array?

Reply:
If custom root is already set (`root !== null`), it empties the data.translations array and writes to the custom root. If custom root is not set, it saves the data array normally as before.

Reply:
What I meant:

```php
		$data = [
			...$existingData,
			'code'         => $this->code(),
			'default'      => $this->isDefault(),
			'direction'    => $this->direction(),
			'locale'       => Locale::export($this->locale()),
			'name'         => $this->name(),
			'url'          => $this->url,
		];

		ksort($data);

		// save translations to the custom root and remove translations
		// to prevent duplication write into the language file
		if ($this->translations()->root() !== null) {
			$this->translations()->save();
		} else {
			$data['translations'] = $this->translations()->toArray();
		}
```

(not passing anything to `$this->translations()->save()` here based on the other suggestion I make below)

Reply:
We need to empty to `$data['translations']` index if custom root defined. Because after few lines later will run `Data::write()` and same translation strings to be written to language translations index. `$data['translations'] = []` prevents it.

Reply:
But if we don't add it to `$data` in the first place, we wouldn't need to unset it, right? Or could `$existingData` also have a `translations` key?

Reply:
Yes, `$existingData` mean `/site/languages/*.php` data and currently languages have `translations` key.